features/256-auth-gear-interaction-overview-in-next/api.md

```diff
@@ -0,0 +1,395 @@
+# auth gear and Cloud Function interaction overview in skygear next
+
+Base on the new product architecture decision (auth gear + Cloud Function and drop record gear), skygear next should bring up a new experience on how auth gear is used in Cloud Function and Client SDK.
+
+## Glossary
+
+* auth gear: a skygear provided component which utilizes user authentication and authorization process, and user profile handling. It would bring enhanced features in the future, such as: JWT provider, Auth UIKit, user management dashboard.
+* Cloud Function: a developer would create a Cloud Function to fulfill application requirements. A cloud function should be a single purpose that attached to certain events or triggered by requirement.
+* auth data: auth related state, such as disabled, last login at, ..., etc.
+* user metaData: indicated as common user properties, such as avatar, first name, last name, display name, ..., etc. 
```

Comment:
I would like to keep "auth data" and "user metaData" idea separately, so I replace the term "user metaData" with "user common properties" because their behavior are different fundamentally.

Reply:
@kaojohnny how are they different? For example, how `disabled` different from `avater` fundamentally, and what are you trying to communicate with the user / other skygear developers when you have two terms/category for them? Asking coz I don't see any differences.

Reply:
@chpapa  `disabled` is related to auth state, which is only changed by specific API (`/auth/disable/set`). 

For `avatar`, `display_name`, `first_name`, they can be changed in a batch (`update_user`, the API doesn't exist yet), and they are not related to auth state. A developer can change them freely.

And second, auth gear allows a hook to change user common attributes and user profile in before hooks, but not support to change auth data in all hooks. Which means, a developer can't simply write `user.disabled = true;` in CF to disable a user. Such action must go through public auth gear API.

Reply:
I think we need a better definition of auth data. Anyway, I will comment on the whole doc again for a better bigger picture to confirm this document.